### DISCUSSION TITLE
Cannot track repositories: `No applicable method: forge-get-url, nil`

### DISCUSSION THREAD

**Discussion by mbollmann** [magit/forge#689] [Support] [answered]
Long-time Magit user, wanted to try Forge for the first time. I followed the steps in the manual, but when calling `N / a` from within Magit (or running `M-x forge-add-repository`) on any repo, I just get the error `No applicable method: forge-get-url, nil`.

I deleted and re-installed Magit, Transient, and Forge, and all of them are at their newest versions:

- `M-x magit-version` gives Magit v4.0.0, Transient 0.7.4, Git 2.46.0, Emacs 29.4, gnu/linux
- Forge is installed at v0.4.0

I’m a bit at a loss why this error might be occuring or how to debug this. Any ideas?

**Comment by tarsius** [comment]
> I’m a bit at a loss why this error might be occuring or how to debug this. Any ideas?

`M-x toggle-debug-on-error` and trigger the issue again. The resulting backtrace will likely contain a bit too much byte-code. Visit the relevant files, likely `forge-commands.el` and `forge-repo.el` and do `M-x eval-buffer`. Now you should get a better backtrace. The cause is likely that Forge doesn't handle some "edge-case" and the backtrace will help me figure out what case that is. (It might not be a very edgy case, the relevant code has seen some heavy changes, not too long ago.)

**Reply by mbollmann** [reply]
Looking at the backtrace and seeing the line `transient-setup(forge-add-repository nil nil :scope nil)` gave me an idea, and I found the root cause for this issue in my own config:

The URLs in my `.git/config` files all start with just `github:`, because I’ve aliased that to `git@github.com` in my `.ssh/config`. It appears that as a result, Forge doesn’t find any repository URLs to add and all the arguments to `forge-add-repository` and `forge-get-url` are `nil`. If I change my git config to say `git@github.com`, it works. I guess it makes sense that this is required, as Forge wouldn’t know otherwise which API it needs to use?

Maybe Forge could handle this error more gracefully, but at least this edge case was on the user’s side :)

**Reply by tarsius** [reply, marked as answer]
Forge was already supposed to resolve ssh host aliases. I've fixed that now (2f5e03f18d604806dd6be090e419b11bceffdece).

If the host cannot be detected or outside a git repository, we still get an error. I'll fix that and probably create a new release tomorrow.

**Reply by tarsius** [reply]
Release is out!

**Reply by mbollmann** [reply]
Awesome, thank you!
